src/raftstore/store/peer.rs

```diff
@@ -749,18 +758,48 @@ impl Peer {
     }
 
     fn handle_raft_commit_entries(&mut self,
-                                  committed_entries: &[eraftpb::Entry])
-                                  -> Result<Vec<ExecResult>> {
+                                  committed_entries: &[eraftpb::Entry],
+                                  mut apply_wb: &mut WriteBatch)
+                                  -> Result<(u64, Vec<ExecResult>)> {
         // If we send multiple ConfChange commands, only first one will be proposed correctly,
         // others will be saved as a normal entry with no data, so we must re-propose these
         // commands again.
         let t = SlowTimer::new();
         let mut results = vec![];
         let committed_count = committed_entries.len();
+        let mut apply_index: u64 = 0;
+        let mut term: u64 = 0;
         for entry in committed_entries {
+            // Entries with different term need apply in different on_raft_ready round.
+            // We call callback before write apply_wb to engine for performance reason,
+            // think about the situation as follows:
+            //   raft group A consists of node 1, 2, 3 and we assume node 1 is leader at
+            //   beginning, when node 1 applied a put command on key "a", if we response to
+            //   client(call cb), it means that if client raises a get "a" command after the
+            //   response of put command, we must return the value we just put, if there
+            //   is no leader transfer happened, we can guarantee this, because the get
+            //   command is run after store.on_raft_ready finished regardless of if it is
+            //   local read or not, but if there is a leader transfer happened, we will resend
+            //   a get command to the new leader(assume node 2), the get command will use raft
+            //   read if the applied index's term is not equal to the read command's term,
+            //   and the front put command and the get command may be applied in the same
+            //   on_raft_ready round in node 2, and the get command may get nothing because
+            //   we have not written the value to engine, so we need to apply the put command
+            //   and get command in different on_raft_ready round.
```

Comment:
Please refer to the following text and update this comment:
The callback is called before writing apply_wb to the engine to improve the performance.
Entries with different terms need to be applied in different on_raft_ready rounds. For more details, see the following example:
The raft group A consists of three nodes: Node 1, Node 2, and Node 3. Node 1 is the leader at the beginning. 
If there is no leader transfer, the process is:
1. Node 1 applies the put command on the "a" key.
2. Node 1 sends a response to a client (call cb).
3. The client raises a get "a" command.
4. Node 1 returns the value ("a") from the put command to the client.
The client can get the value from the put command because the get command is run after store.on_raft_ready is finished.
However, if the leader is transferred from Node 1 to Node 2, the process is:
1. Node 1 applies the put command on the "a" key.
2. Node 1 sends a response to a client (call cb).
3. The client sends a get command to the new leader (Node 2).
4. The get command uses Raft read because of the leader transfer. 
In this case, the put command and the get command from the client may be applied in the same on_raft_ready round on Node 2. Because the "a" value from the put command is not written to the engine, the get command cannot get the value ("a"). So the put command and get command must be appiled in different on_raft_ready rounds.